scripts/generate-clients.mjs

```diff
@@ -7,66 +7,95 @@ import { renderVisitor as renderRustVisitor } from "@kinobi-so/renderers-rust";
 import { getAllProgramIdls } from "./utils.mjs";
 
 // Instanciate Kinobi.
-const [idl, ...additionalIdls] = getAllProgramIdls().map(idl => rootNodeFromAnchor(require(idl)))
+const [idl, ...additionalIdls] = getAllProgramIdls().map((idl) =>
+  rootNodeFromAnchor(require(idl))
+);
 const kinobi = k.createFromRoot(idl, additionalIdls);
 
 // Update programs.
 kinobi.update(
   k.updateProgramsVisitor({
-    "stakeProgram": { name: "stake" },
+    stakeProgram: { name: "stake" },
   })
 );
 
-// Update accounts.
+// Add PDA information.
 kinobi.update(
-  k.updateAccountsVisitor({
-    counter: {
-      seeds: [
-        k.constantPdaSeedNodeFromString("utf8", "counter"),
-        k.variablePdaSeedNode(
-          "authority",
-          k.publicKeyTypeNode(),
-          "The authority of the counter account"
-        ),
-      ],
+  k.bottomUpTransformerVisitor([
+    {
+      select: "[programNode]stake",
+      transform: (node) => {
+        k.assertIsNode(node, "programNode");
+        return {
+          ...node,
+          pdas: [
+            k.pdaNode({
+              name: "vault",
+              seeds: [
+                k.constantPdaSeedNodeFromString("utf8", "token-vault"),
+                k.variablePdaSeedNode(
+                  "authority",
+                  k.publicKeyTypeNode(),
+                  "Config authority"
+                ),
+              ],
+            }),
+          ],
+        };
+      },
```

Comment:
I know Kinobi isn't fully documented yet, so it would be great to have some context here on what this is doing.

From what I can gather, it seems like it's declaring the `vault` field as a PDA, and the seeds are the string literal `"token-vault"` and the public key in the `authority` field?

I'm not sure on the other bits like `select: "[programNode]stake",`...

Reply:
This is declaring what usually is referred to as a "ghost" PDA, but I think a better name is a "signer" PDA. Kinobi has the option to configure PDAs without necessarily being an account on the program.

Everything in Kinobi is a node, so we need to select the program node, then access its `pdas` property (all program nodes have a `pdas` property), then injected a new PDA.

In this case, we are saying that the program `stake`, selected with the `select: "[programNode]stake"` expression, has a PDA named `vault`; the seeds are as you described, but the `authority` is not a field but a parameter. This will generate a helper `findVaultPda` that will take an "authority" address as a seed – the constant `"token-vault"` is hard-coded so there is no need to pass it as a parameter.

PS: The helper so far only exists on the JS side but I am about to add it to the Rust client as well.

Reply:
Cool, thanks for the explanation! Makes sense, and sounds useful the client!

If the program needs to use this PDA's derivation - say in a helper - to validate some checks in the processor, does that mean we'll have to redefine it in the program as well? Like this, for example:
https://github.com/paladin-bladesmith/rewards-program/blob/3bf1c13c2f619736c1afb759054e71e19eb55c59/program/src/state.rs#L18-L21

Reply:
We can have helpers generated for the program using Shank's `seeds` macro: https://github.com/metaplex-foundation/shank/blob/master/shank-macro/src/lib.rs#L34-L71

I think we talked about this and I might have given you the answer that you couldn't, but that is for the client.

So currently, `seeds` is not going to get exported on the IDL so you need to add the PDA information twice: once with `seeds` and another on your Kinobi config.

This will change once we have Kinobi specific macros.